dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/TimeToLive.java

```diff
@@ -0,0 +1,116 @@
+import static java.time.ZoneOffset.UTC;
+import static java.time.temporal.ChronoUnit.DAYS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_FACTOR;
+import static org.springframework.util.Assert.notNull;
+
+import java.time.Instant;
+
+import org.hisp.dhis.analytics.DataQueryParams;
+import org.hisp.dhis.setting.SystemSettingManager;
+
+public class TimeToLive
+    implements
+    Computable
+{
+
+    static final int DEFAULT_TTL_FACTOR = 5;
+    static final long DEFAULT_MULTIPLIER = 1;
+
+    private final DataQueryParams params;
+
+    private SystemSettingManager systemSettingManager;
+
+    public TimeToLive( final DataQueryParams params, final SystemSettingManager systemSettingManager )
+    {
+        notNull( params, "Object params must not be null" );
+        notNull( params.getLatestEndDate(), "Object params.getLatestEndDate() must not be null" );
+        notNull( systemSettingManager, "Object systemSettingManager must not be null" );
+        this.params = params;
+        this.systemSettingManager = systemSettingManager;
+    }
+
+    /**
+     * Execute the internal rules in order to calculate a TTL for the given
+     * parameters. The current rules are based on a configurable timeout
+     * "factor" (through SettingKey) which will be used in the calculation
+     * of this time to live. Given the "factor" described above:
+     *
+     * Older the "endingDate", higher the "factor", longer the TTL.
+     * The formula is basically: TTL = "factor" * (diff between now and endingDate)
+     *
+     * @return the computed TTL value in SECONDS.
+     */
+    @Override
+    public long compute()
+    {
+        final Instant endingDate = params.getLatestEndDate().toInstant();
+
+        /*
+         * If the difference between the most recent date and NOW is 0 (zero) it means
+         * the current day, so we increment the multiplier by 1 (one) avoiding
+         * multiplying by 0 (zero).
+         */
+        final long diff = daysBetweenDateAndNow( endingDate );
+        final long ttlMultiplier = diff > 0 ? diff : DEFAULT_MULTIPLIER;
+
+        return ttlFactorOrDefault() * ttlMultiplier;
+    }
+
+    /**
+     * Calculates the difference between now and the given date.
+     * It has a the particularity of returning ZERO (0), if the
+     * diff is negative, which means the date is ahead of now.
+     *
+     * @param date the date to subtract from now
+     * @return the difference of days in MILLISECONDS
+     */
+    private long daysBetweenDateAndNow( final Instant date )
+    {
+        final long diff = DAYS.between( ofInstant( date, UTC ), now() );
+        return diff >= 0 ? diff : 0;
+    }
+
+    /**
+     * Returns the default TTL factor or a default one if none is defined.
+     * @return the factor in
+     */
+    private int ttlFactorOrDefault()
+    {
+        final Integer ttlFactor = (Integer) systemSettingManager.getSystemSetting( ANALYTICS_CACHE_FACTOR );
```

Comment:
Here we can just utilize the default value coming from the system setting `ANALYTICS_CACHE_FACTOR`. Since we have specified a default value for ANALYTICS_CACHE_FACTOR, it can never return null / will return the default if not set. That also allows us to remove `DEFAULT_TTL_FACTOR` and centralize the value (avoid having two).

We do need to validate that the value is not negative as you say and that should ideally happen in the API but not sure if we actually do that. We could have a check to say that if the value is < 1, then use the system setting default value.

Reply:
Done.

Currently, we don't have validations when saving a system setting variable.
My suggestion it to create a generic validation as part of other task.